python/lsst/ts/dcs/mock_controller.py

```diff
@@ -0,0 +1,249 @@
+            port,
+    ):
+        self.port = port
+        self._server = None
+        self.log = logging.getLogger("MockDomeController")
+        # Dict of command: (has_argument, function).
+        # The function is called with:
+        # * No arguments, if `has_argument` False.
+        # * The argument as a string, if `has_argument` is True.
+        self.dispatch_dict = {
+            "status": (False, self.status),
+            "Dome_command_moveAz": (False, self.dome_command_move_az),
+            "Dome_command_moveEl": (False, self.dome_command_move_el),
+            "Dome_command_stopAz": (False, self.dome_command_stop_az),
+            "Dome_command_stopEl": (False, self.dome_command_stop_el),
+            "quit": (False, self.quit),
+        }
+        # Name of a command to report as failed once, the next time it is seen,
+        # or None if no failures. Used to test CSC handling of failed commands.
+        self.fail_command = None
+
+    async def start(self):
+        """Start the TCP/IP server.
+
+        Set start_task done and start the command loop.
+        """
+        self.log.info("start called")
+        self._server = await asyncio.start_server(
+            self.cmd_loop, host="127.0.0.1", port=self.port
+        )
+        await self._server.serve_forever()
+
+    async def stop(self):
+        """Stop the TCP/IP server.
+        """
+        if self._server is None:
+            return
+
+        server = self._server
+        self._server = None
+        logging.info("closing server")
+        server.close()
+        logging.info("done closing")
+
+    async def cmd_loop(self, reader, writer):
+        """Execute commands and output replies."""
+        self.log.info("cmd_loop begins")
+        writer.write("Enter Command > ".encode())
+        await writer.drain()
+        while True:
+            line = await reader.readline()
+            line = line.decode()
+            if not line:
+                # connection lost; close the writer and exit the loop
+                self.log.info("closing the writer")
+                writer.close()
+                self.log.info("stopping mock controller")
+                await self.stop()
+                return
+            line = line.strip()
+            self.log.info(f"read command: {line!r}")
+            if line:
+                try:
+                    items = line.split()
+                    cmd = items[-1]
+                    if cmd not in self.dispatch_dict:
+                        raise KeyError(f"Unsupported command {cmd}")
+                    if cmd == self.fail_command:
+                        self.fail_command = None
+                        outputs = [f"Command {cmd} failed by request"]
+                    else:
+                        has_data, func = self.dispatch_dict[cmd]
+                        desired_len = 2 if has_data else 1
+                        if len(items) != desired_len:
+                            raise RuntimeError(
```

Comment:
Same comment as in the previous case. Instead of raising RuntimeError, I think it is better to respond with an error message.

Reply:
Well, I copied that from the mock_controller code in ts_ATDome ;-) However, I agree that returning a failure code would be better. I'll need to discuss this with German and EIE to see how we solve this. Once we decide how I will open a new ticket to solve this and the other Exception as well.